### PR TITLE
Fix keyword argument warning on Action Mailer

### PR DESCRIPTION
**Pull request by rafaelfranca**

### CODE REVIEW

File: actionmailer/lib/action_mailer/message_delivery.rb [hunk 1]
```diff
@@ -143,9 +143,9 @@ def enqueue_delivery(delivery_method, options = {})
 
       def arguments_for(delivery_job, delivery_method)
         if delivery_job <= MailDeliveryJob
-          [@mailer_class.name, @action.to_s, delivery_method.to_s, args: @args]
+          [@mailer_class.name, @action.to_s, delivery_method.to_s, args: @args, kwargs: @kwargs]
```
Comment (rafaelfranca) [inline actionmailer/lib/action_mailer/message_delivery.rb]: This change is a breaking change and I saw no way to make this work without a breaking change. Applications upgrading will lost jobs because either the job worker may be using the old code and not accepting this new argument. We have the same problem in the `else` branch.

Comment (rafaelfranca) [reply actionmailer/lib/action_mailer/message_delivery.rb]: @eregon @mame ideas on how to avoid making this braking change to applications while still removing this warning?

Comment (eregon) [reply actionmailer/lib/action_mailer/message_delivery.rb]: I suspect @jeremyevans and @mame's answer would be something like:

Use `ruby2_keywords`, and only use `*args, **kwargs` when dropping Ruby 2.x support, since that doesn't delegate correctly in 2.6 (`**kwargs` passes an extra positional `{}` if no kwargs passed) and in 2.7 (`delegate({})` drops positional `{}`).

Comment (rafaelfranca) [reply actionmailer/lib/action_mailer/message_delivery.rb]: This would not fix the warning, unless application developers call `ruby2_keywords` in all the actions of their mailers. I don't think application developers should call that method since this is implementation detail of the framework.

Comment (rafaelfranca) [reply actionmailer/lib/action_mailer/message_delivery.rb]: Here is the changeset with `ruby2_keywords` https://github.com/rails/rails/compare/a55fa7911a. And the warning will be found on this CI job https://buildkite.com/rails/rails/builds/65996#ef0e092f-3e67-46da-8fba-fc7587e63bd8

Comment (jeremyevans) [reply actionmailer/lib/action_mailer/message_delivery.rb]: Can you please either provide an minimal, self contained reproducible example showing the issue in the failed CI job ( https://buildkite.com/rails/rails/builds/65996#ef0e092f-3e67-46da-8fba-fc7587e63bd8), or at least the complete call stack (`puts caller` inside `bounce` should work).  It's possible the keywords are converted to a positional hash earlier, before `process` is called, in which case the warning is correct, and there is another place `ruby2_keywords` needs to be added.  You should only need `ruby2_keywords` in user code if the user code is what is doing the delegating of arguments.

Comment (eregon) [reply actionmailer/lib/action_mailer/message_delivery.rb]: I think https://bugs.ruby-lang.org/issues/16463 might be a way to fix warnings without changing any delegation code.
In any case, it seems worth finding out why `ruby2_keywords` is not enough here, or if it's needed in user code.

Comment (mame) [reply actionmailer/lib/action_mailer/message_delivery.rb]: Since I'm unfamiliar with Rails, I can't yet understand what is a problem precisely.  As long as I see only this file, saving `@kwargs` is not needed, and adding only `ruby2_keywords :initialize` is good enough (I didn't test it, sorry).  Or, do many users inherit this MessageDelivery class and override `initialize` method?  Minimal example will be helpful to me, too.

Comment (rafaelfranca) [reply actionmailer/lib/action_mailer/message_delivery.rb]: Here is a minimal script with the current behavior.

```ruby
# This is framework code
class MailDelivery
  def initialize(mailer_class, action, *args)
    @mailer_class = mailer_class
    @action = action
    @args = args
  end

  def deliver_now
    @mailer_class.new.process @action, *@args
  end
end

class RailsMailer
  class << self
    def method_missing(method_name, *args)
      MailDelivery.new(self, method_name, *args)
    end
  end

  def process(method_name, *args)
    process_action(method_name, *args)
  end

  def process_action(method_name, *args)
    send_action(method_name, *args)
  end

  alias send_action send
end
# End of framework code

# This is users code
class UserMailer < RailsMailer
  def bounce(to:)
    puts to
  end
end

UserMailer.bounce(to: "rafael").deliver_now
# End of user code
```

Comment (eregon) [reply actionmailer/lib/action_mailer/message_delivery.rb]: This is a way to fix it with `ruby2_keywords`, without changing user code:
```ruby
# This is framework code
class MailDelivery
  ruby2_keywords def initialize(mailer_class, action, *args)
    @mailer_class = mailer_class
    @action = action
    @args = args
  end

  def deliver_now
    @mailer_class.new.process @action, *@args
  end
end

class RailsMailer
  class << self
    ruby2_keywords def method_missing(method_name, *args)
      MailDelivery.new(self, method_name, *args)
    end
  end

  def process(method_name, *args)
    process_action(method_name, *args)
  end

  def process_action(method_name, *args)
    send_action(method_name, *args)
  end

  alias send_action send
end
# End of framework code

# This is users code
class UserMailer < RailsMailer
  def bounce(to:)
    puts to
  end
end

UserMailer.bounce(to: "rafael").deliver_now
# End of user code
```

The warning, which is:
```
mail.rb:26: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
mail.rb:35: warning: The called method `bounce' is defined here
```
is really unhelpful here, both lines are far from where `ruby2_keywords` needs to be added.
`puts caller` gives:
```
mail.rb:26:in `process_action'
mail.rb:22:in `process'
mail.rb:10:in `deliver_now'
mail.rb:41:in `<main>'
```
And `deliver_now` is the relevant method actually delegating here, so it needs `@args` to be flagged by `ruby2_keywords`, in `initialize`.
That's not enough, because `method_missing` also needs it, otherwise it "forgets" that it was passed keyword arguments (doesn't flag the passed keyword arguments Hash).

Comment (eregon) [reply actionmailer/lib/action_mailer/message_delivery.rb]: In other words one needs to basically follow every call and understand the whole chain to have a chance to find out where to add `ruby2_keywords`. I'd rather `ruby2_keywords` to be the default and not get such headaches for little value: https://bugs.ruby-lang.org/issues/16463

Comment (eregon) [reply actionmailer/lib/action_mailer/message_delivery.rb]: @jeremyevans @mame Why do we need `ruby2_keywords` on `initialize`, shouldn't just on `method_missing` (the first method called by user code) be enough, since that flags the Hash as kwargs? There is no copy of the Hash here.
Conversely, why do `process` and `process_action` don't need `ruby2_keywords`? I think at some point Jeremy wanted the kwargs flag to be removed after a `*args` call.

Comment (jeremyevans) [reply actionmailer/lib/action_mailer/message_delivery.rb]: Every time you splat an array where the final argument is a keyword-flagged hash (e.g. `MailDelivery.new(self, method_name, *args)`), the keyword-flagged hash is duped and the keyword flag is removed.  If that method needs to treat the final argument as keywords in some later call (e.g. `@mailer_class.new.process @action, *@args`), then it needs to be flagged with `ruby2_keywords` so the duped hash has the keyword flag added to it (e.g. `ruby2_keywords` is needed on `MailDelivery#initialize`).

You do not need `ruby2_keywords` on `process` or `process_action` because the arguments they receive are not used in a method call that would treat the final argument as keywords.

In short, there is no bug in Ruby here.  The warning emitted does not directly point to the error, which is unfortunate, but looking at the call stack it is possible to determine where `ruby2_keywords` should be introduced.

Comment (eregon) [reply actionmailer/lib/action_mailer/message_delivery.rb]: > You do not need `ruby2_keywords` on `process` or `process_action` because the arguments they receive are not used in a method call that would treat the final argument as keywords.

That part I don't get. The call chain is:
```
deliver_now -> process -> process_action -> send -> UserMailer#bounce
```
And that last method does need the last argument as keywords, isn't it?
Why do `deliver_now`, `process` and `process_action` not lose the flag when the `MailDelivery.new(self, method_name, *args)` call does lose it?

Comment (eregon) [reply actionmailer/lib/action_mailer/message_delivery.rb]: When trying to debug it in https://gist.github.com/eregon/af26c29506b3de640bab1a1d15d228bf I see that the Hash gets dup'd for `MailDelivery.new(self, method_name, *args)`, but not for all other calls with `*args`.
Is that an implementation detail of `Class#new` leaking here? (or how `initialize` is called from it maybe?)

What if `Class#new` was implemented in Ruby?
Actually that removes the need for an extra `ruby2_keywords` on `initialize`!

Implementing it as `allocate.tap { |obj| obj.send(:initialize, *args) }` seems to also dup the Hash, (a bug of `send` it seems?) but making `initialize` methods `public` and using `allocate.tap { |obj| obj.initialize(*args) }` does **not** dup the Hash, allowing the whole example to work with `ruby2_keywords` only on the `method_missing` (first call receiving kwargs):
https://gist.github.com/eregon/95f9aba8f5d05ea58e93c1295d97c6f3

Unless I'm missing something obvious, this is extremely complicated to reason about, and highly inconsistent in where `ruby2_keywords` is needed.
IMHO, one more reason to enable it for all methods by default (https://bugs.ruby-lang.org/issues/16463), which at the same time avoids confusing warnings and does not require to guess (and get headaches) where to add `ruby2_keywords` (which is also not so nice in the code).

Comment (mame) [reply actionmailer/lib/action_mailer/message_delivery.rb]: @rafaelfranca Thank you very much for the minimal code!  It is very helpful for me to understand the issue.

@eregon 

> Why do `deliver_now`, `process` and `process_action` not lose the flag when the MailDelivery.new(self, method_name, *args) call does lose it?

Agreed.  I think this is a false negative.  `process` and `process_action` should be also marked as `ruby2_keywords`.  I've created https://bugs.ruby-lang.org/issues/16466.

Comment (eregon) [reply actionmailer/lib/action_mailer/message_delivery.rb]: I also tried to make the real test work without warnings, which is much more complicated than this minimal example.
I couldn't find a solution after hours of debugging it and adding `ruby2_keywords` in 8 places.
I wonder who can.
BTW, a way to detect if a Hash is flagged as keyword arguments would be useful for debugging.

To run the test:
```
git clone rails
cd rails/actionmailbox
bundle install
bundle exec rake test TEST=test/unit/mailbox/bouncing_test.rb
```

The challenge is to fix this warning:
```
rails/actionpack/lib/abstract_controller/base.rb:195: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
rails/actionmailbox/test/test_helper.rb:51: warning: The called method `bounce' is defined here
```

Comment (eregon) [reply actionmailer/lib/action_mailer/message_delivery.rb]: After a lot of debugging, with [my branch enabling `ruby2_keywords`  by default](https://github.com/ruby/ruby/compare/master...eregon:ruby2_keywords-by-default) and the help of `Hash#flagged?` I found that [SPOILER] it's related to serialization and deserialization in https://github.com/rails/rails/blob/09cc7967/activejob/lib/active_job/arguments.rb
That effectively loses the "keyword arguments" flag.

It's not to difficult to remember that, similar to how symbols keys are handled there, but we need a way to find out if a Hash is flagged, and a way to flag a Hash.
In my case I used:
```ruby
private def flag_hash(*args) # would need ruby2_keywords if not using my branch
  args[-1]
end

is_flagged = argument.flagged? # serialize, Hash#flagged? is not in 2.7.0
...
result = flag_hash(**result) if is_flagged # deserialize
```
But obviously we'd want something more direct.

Comment (eregon) [reply actionmailer/lib/action_mailer/message_delivery.rb]: Remembering the flag during serialization, together with ruby2_keywords by default, actually fixes the warning.

Comment (mame) [reply actionmailer/lib/action_mailer/message_delivery.rb]: Solved :-)

https://gist.github.com/mame/18cd2dfc7d965d0bad1216f2fdd008ee

I admit that it is super hacky, though.

ActiveJob serializes and deserializes the argument objects, which removes the `ruby2_keywords` flag.  This patch preserves the flag explicitly.

This may indicate the need of a feature to inspect and add the `ruby2_keywords` flag to a hash.  @jeremyevans What do you think?

Comment (eregon) [reply actionmailer/lib/action_mailer/message_delivery.rb]: Congrats, almost at the same time than me :)

This is the full patch with ruby2_keywords by default for comparison:
https://gist.github.com/eregon/cd60d95fd7c448379e94afb21e2f8a7c

There is no need to find the 8 methods that might or not need `ruby2_keywords`.

**Comment by kamipo** [comment]
This has fixed by #38260.

**Comment by rafaelfranca** [comment]
Wait, what? That fix is not acceptable. Can we revert? We should not be increasing the storage on redis or any other system because Ruby is showing warnings.

**Comment by kamipo** [comment]
Reverted 99f18b618b990fd271fcb6f6beb0052dd55064f9.

**Comment by rafaelfranca** [comment]
#38260 has most of what we need, so let's reopen it. We need to make the code future proof so we don't break again when we drop support to `ruby2_keywords`.

We really need to discuss if it is really the right thing to do to ask users to store more data on they datastores to remove a warning in the language

I don't think this makes sense, and this unfortunately was a case that Ruby core missed when making this decision. If you serialize method arguments anywhere you will have to increase the serialization storage and some systems can't easily make room for this new space.

Can we solve this problem with a version check? First, make the job work with the arguments split or not. Then make another version that always split on Ruby 2.7 or don't split on Ruby 2.6.
